### ISSUE TITLE
Occasional "broken pipe" error in ember server

### ISSUE THREAD

**Issue by retriku** [http4s/http4s#6219]
Occasionally, I am getting following error when running `wrk` on http4s server with ember backend
```
java.io.IOException: Broken pipe
	at java.base/sun.nio.ch.FileDispatcherImpl.write0(Native Method)
	at java.base/sun.nio.ch.SocketDispatcher.write(Unknown Source)
	at java.base/sun.nio.ch.IOUtil.writeFromNativeBuffer(Unknown Source)
	at java.base/sun.nio.ch.IOUtil.write(Unknown Source)
	at java.base/sun.nio.ch.IOUtil.write(Unknown Source)
	at java.base/sun.nio.ch.UnixAsynchronousSocketChannelImpl.finishWrite(Unknown Source)
	at java.base/sun.nio.ch.UnixAsynchronousSocketChannelImpl.finish(Unknown Source)
	at java.base/sun.nio.ch.UnixAsynchronousSocketChannelImpl.onEvent(Unknown Source)
	at java.base/sun.nio.ch.EPollPort$EventHandlerTask.run(Unknown Source)
	at java.base/sun.nio.ch.AsynchronousChannelGroupImpl$1.run(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
	at java.base/java.lang.Thread.run(Unknown Source)
	at delay @ fs2.io.net.SocketCompanionPlatform$AsyncSocket.$anonfun$write$1(SocketPlatform.scala:137)
	at async @ fs2.io.net.SocketCompanionPlatform$AsyncSocket.go$2(SocketPlatform.scala:131)
	at flatMap @ fs2.io.net.SocketCompanionPlatform$AsyncSocket.go$2(SocketPlatform.scala:141)
	at modify @ fs2.internal.Scope.close(Scope.scala:262)
	at flatMap @ fs2.Compiler$Target.flatMap(Compiler.scala:162)
	at deferred @ fs2.concurrent.Topic$.apply(Topic.scala:144)
	at acquire @ fs2.Stream.go$11(Stream.scala:1922)
```

Setup:
* Scala 2.13.8
* Cats Effect 3.3.8
* Fs2 3.2.5
* Http4s 0.23.11

Mentioned error is not logged by logger and is thrown time to time (seems randomly). I wonder what could have caused it? Setup of process running `wrk` is as follows, wrk opens one connection to http4s server, and 100 instances of wrk script are run in parallel on same linux container. Each wrk process connects to different url path on same server.

**Comment by FloWi** [comment]
I have the same issue. For me it happens when my scala-js frontend creates a `<video>` tag with its src pointing to a static url that is handled by http4s.

I created the endpoint like this...
```scala
  val httpApp: Kleisli[IO, Request[IO], Response[IO]] =
    Router(
      "api"    -> apiService,
      "assets" -> Logger.httpRoutes(logHeaders = true, logBody = false)(
        fileService[IO]("assets")),
      ),
    ).orNotFound
```

...and the embedded video like this.
```html
<video src="http://localhost:8080/assets/recordings/week%201/02%20Supervised%20learning%20part%201/index.mp4#t=63" 
  controls="true"
  type="video/mp4" 
  style="width: 90%;">
</video>
```

These are the two requests and below are the errors.
I noticed that those errors are written directly to stderr and are not captured by logback.
Fun fact - I tried to reproduce the log-to-stderr behavior in a minimal-project and there the error is silently handled. I have no idea why.

I'm able to capture the error like this, but I couldn't find a way to silence it without redirecting stderr to /dev/null. It's a bit annoying, since the log shows plenty of errors but everything works fine and the browser doesn't show any error.
```scala
      .withErrorHandler { case th: Throwable =>
        logger.error(th)("Error") *> InternalServerError("Houston, we have a problem " + th.getMessage)
      }
```

Does anyone have an idea what causes the app to log to stderr (and doesn't in another app that's almost identical)?

```
[io-compute-2] INFO  o.h.s.m.Logger - HTTP/1.1 GET /assets/recordings/week%201/02%20Supervised%20learning%20part%201/index.mp4 Headers(Host: localhost:8080, Connection: keep-alive, Pragma: no-cache, Cache-Control: no-cache, sec-ch-ua: ".Not/A)Brand";v="99", "Google Chrome";v="103", "Chromium";v="103", Accept-Encoding: identity;q=1, *;q=0, sec-ch-ua-mobile: ?0, User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/103.0.0.0 Safari/537.36, sec-ch-ua-platform: "macOS", Accept: */*, Sec-Fetch-Site: same-site, Sec-Fetch-Mode: no-cors, Sec-Fetch-Dest: video, Referer: http://localhost:12345/, Accept-Language: de-DE,de;q=0.9,en-US;q=0.8,en;q=0.7, Cookie: <REDACTED>, Range: bytes=0-) 
[io-compute-2] INFO  o.h.s.m.Logger - HTTP/1.1 206 Partial Content Headers(Last-Modified: Thu, 30 Jun 2022 14:11:32 GMT, Content-Length: 14541784, Content-Type: video/mp4, ETag: "181b4f3e854-dde3d8", Accept-Ranges: bytes, Content-Range: bytes 0-14541783/14541784) 

[io-compute-13] INFO  o.h.s.m.Logger - HTTP/1.1 GET /assets/recordings/week%201/02%20Supervised%20learning%20part%201/index.mp4 Headers(Host: localhost:8080, Connection: keep-alive, Pragma: no-cache, Cache-Control: no-cache, sec-ch-ua: ".Not/A)Brand";v="99", "Google Chrome";v="103", "Chromium";v="103", Accept-Encoding: identity;q=1, *;q=0, sec-ch-ua-mobile: ?0, User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/103.0.0.0 Safari/537.36, sec-ch-ua-platform: "macOS", Accept: */*, Sec-Fetch-Site: same-site, Sec-Fetch-Mode: no-cors, Sec-Fetch-Dest: video, Referer: http://localhost:12345/, Accept-Language: de-DE,de;q=0.9,en-US;q=0.8,en;q=0.7, Cookie: <REDACTED>, Range: bytes=2916352-) 
[io-compute-13] INFO  o.h.s.m.Logger - HTTP/1.1 206 Partial Content Headers(Last-Modified: Thu, 30 Jun 2022 14:11:32 GMT, Content-Length: 11625432, Content-Type: video/mp4, ETag: "181b4f3e854-dde3d8", Accept-Ranges: bytes, Content-Range: bytes 2916352-14541783/14541784) 
````

```text
java.io.IOException: Broken pipe
	at java.base/sun.nio.ch.FileDispatcherImpl.write0(Native Method)
	at java.base/sun.nio.ch.SocketDispatcher.write(SocketDispatcher.java:47)
	at java.base/sun.nio.ch.IOUtil.writeFromNativeBuffer(IOUtil.java:113)
	at java.base/sun.nio.ch.IOUtil.write(IOUtil.java:79)
	at java.base/sun.nio.ch.IOUtil.write(IOUtil.java:50)
	at java.base/sun.nio.ch.UnixAsynchronousSocketChannelImpl.finishWrite(UnixAsynchronousSocketChannelImpl.java:591)
	at java.base/sun.nio.ch.UnixAsynchronousSocketChannelImpl.finish(UnixAsynchronousSocketChannelImpl.java:195)
	at java.base/sun.nio.ch.UnixAsynchronousSocketChannelImpl.onEvent(UnixAsynchronousSocketChannelImpl.java:213)
	at java.base/sun.nio.ch.KQueuePort$EventHandlerTask.run(KQueuePort.java:312)
	at java.base/sun.nio.ch.AsynchronousChannelGroupImpl$1.run(AsynchronousChannelGroupImpl.java:112)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
	at delay @ fs2.io.net.SocketCompanionPlatform$AsyncSocket.$anonfun$write$1(SocketPlatform.scala:137)
	at async @ fs2.io.net.SocketCompanionPlatform$AsyncSocket.go$2(SocketPlatform.scala:131)
	at flatMap @ fs2.io.net.SocketCompanionPlatform$AsyncSocket.go$2(SocketPlatform.scala:141)
	at delay @ fs2.io.net.SocketCompanionPlatform$AsyncSocket.$anonfun$write$1(SocketPlatform.scala:137)
	at async @ fs2.io.net.SocketCompanionPlatform$AsyncSocket.go$2(SocketPlatform.scala:131)
	at flatMap @ fs2.io.net.SocketCompanionPlatform$AsyncSocket.go$2(SocketPlatform.scala:141)
	at modify @ org.http4s.ember.server.internal.Shutdown$$anon$1.<init>(Shutdown.scala:83)
java.io.IOException: Connection reset by peer
	at java.base/sun.nio.ch.FileDispatcherImpl.read0(Native Method)
	at java.base/sun.nio.ch.SocketDispatcher.read(SocketDispatcher.java:39)
	at java.base/sun.nio.ch.IOUtil.readIntoNativeBuffer(IOUtil.java:276)
	at java.base/sun.nio.ch.IOUtil.read(IOUtil.java:233)
	at java.base/sun.nio.ch.IOUtil.read(IOUtil.java:223)
	at java.base/sun.nio.ch.UnixAsynchronousSocketChannelImpl.finishRead(UnixAsynchronousSocketChannelImpl.java:395)
	at java.base/sun.nio.ch.UnixAsynchronousSocketChannelImpl.finish(UnixAsynchronousSocketChannelImpl.java:191)
	at java.base/sun.nio.ch.UnixAsynchronousSocketChannelImpl.onEvent(UnixAsynchronousSocketChannelImpl.java:213)
	at java.base/sun.nio.ch.KQueuePort$EventHandlerTask.run(KQueuePort.java:312)
	at java.base/sun.nio.ch.AsynchronousChannelGroupImpl$1.run(AsynchronousChannelGroupImpl.java:112)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
	at delay @ fs2.io.net.SocketCompanionPlatform$AsyncSocket.$anonfun$readChunk$1(SocketPlatform.scala:123)
	at async @ fs2.io.net.SocketCompanionPlatform$AsyncSocket.readChunk(SocketPlatform.scala:117)
	at flatMap @ fs2.io.net.SocketCompanionPlatform$BufferedReads.$anonfun$read$1(SocketPlatform.scala:85)
	at delay @ fs2.io.net.SocketCompanionPlatform$BufferedReads.$anonfun$withReadBuffer$1(SocketPlatform.scala:54)
	at flatten$extension @ org.http4s.ember.server.internal.Shutdown$$anon$1.<init>(Shutdown.scala:71)
	at modify @ org.http4s.ember.server.internal.Shutdown$$anon$1.<init>(Shutdown.scala:83)
	at flatten$extension @ org.http4s.ember.server.internal.Shutdown$$anon$1.<init>(Shutdown.scala:71)
	at apply @ org.http4s.ember.server.EmberServerBuilder.$anonfun$build$2(EmberServerBuilder.scala:179)
java.io.IOException: Broken pipe
	at java.base/sun.nio.ch.FileDispatcherImpl.write0(Native Method)
	at java.base/sun.nio.ch.SocketDispatcher.write(SocketDispatcher.java:47)
	at java.base/sun.nio.ch.IOUtil.writeFromNativeBuffer(IOUtil.java:113)
	at java.base/sun.nio.ch.IOUtil.write(IOUtil.java:79)
	at java.base/sun.nio.ch.IOUtil.write(IOUtil.java:50)
	at java.base/sun.nio.ch.UnixAsynchronousSocketChannelImpl.finishWrite(UnixAsynchronousSocketChannelImpl.java:591)
	at java.base/sun.nio.ch.UnixAsynchronousSocketChannelImpl.finish(UnixAsynchronousSocketChannelImpl.java:195)
	at java.base/sun.nio.ch.UnixAsynchronousSocketChannelImpl.onEvent(UnixAsynchronousSocketChannelImpl.java:213)
	at java.base/sun.nio.ch.KQueuePort$EventHandlerTask.run(KQueuePort.java:312)
	at java.base/sun.nio.ch.AsynchronousChannelGroupImpl$1.run(AsynchronousChannelGroupImpl.java:112)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
	at delay @ fs2.io.net.SocketCompanionPlatform$AsyncSocket.$anonfun$write$1(SocketPlatform.scala:137)
	at async @ fs2.io.net.SocketCompanionPlatform$AsyncSocket.go$2(SocketPlatform.scala:131)
	at flatMap @ fs2.io.net.SocketCompanionPlatform$AsyncSocket.go$2(SocketPlatform.scala:141)
	at modify @ org.http4s.ember.server.internal.Shutdown$$anon$1.<init>(Shutdown.scala:83)
	at flatten$extension @ org.http4s.ember.server.internal.Shutdown$$anon$1.<init>(Shutdown.scala:71)
	at apply @ org.http4s.ember.server.EmberServerBuilder.$anonfun$build$2(EmberServerBuilder.scala:179)
	at acquire @ fs2.Stream$NestedStreamOps$.$anonfun$parJoin$19(Stream.scala:4043)
```

**Comment by armanbilge** [comment]
@FloWi what's your version of http4s and Cats Effect? If things are working normally but an errors is being logged to stderr it could be Cats Effect's new "unhandled error handling" feature.

**Comment by FloWi** [comment]
Lol - when I was checking the versions, I noticed a difference between my two projects. I overlooked that difference when comparing both versions.
The one with the "wrong" behavior was using http4s version 0.23.12 (since I created it last week with your g8 template). The version I created yesterday used 0.23.23 works as expected and doesn't spam stderr 🎉. I saw that Ross upgraded the g8 template _this_ week ;-)

**Comment by FloWi** [comment]
Btw: thanks for working on http4s, @armanbilge and others. I really enjoy working with it. The documentation is awesome compared to other open source projects!

**Comment by armanbilge** [comment]
@retriku do you still have this issue with the latest Cats Effect as well? Seems to me it was the changes in Cats Effect v3.3.8 which caused this behavior and has since been improved. See:
- https://github.com/typelevel/cats-effect/pull/2868
which was released in https://github.com/typelevel/cats-effect/releases/tag/v3.3.8

**Comment by armanbilge** [comment]
Laika reports frequent broken pipe errors when using SSE.
- https://github.com/planet42/Laika/pull/311
